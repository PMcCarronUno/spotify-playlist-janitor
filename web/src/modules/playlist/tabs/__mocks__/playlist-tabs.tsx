import React from 'react';
export const PlaylistTabs = () => <div>Playlist Tabs</div>;
