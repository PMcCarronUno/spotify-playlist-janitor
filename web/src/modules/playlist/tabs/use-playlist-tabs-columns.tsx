import { PiClock } from "react-icons/pi";
import { VscTrash } from "react-icons/vsc";
import format from "date-fns/format";
import { Column } from "shared/components/table";
import { SkippedTrackAlbum, SkippedTrackArtist } from "./playlist-tabs-types";
import { Image } from "shared/types";

type UsePlaylistTabsColumnsProps = {
  onDeleteClick(id: string): void;
};

export const usePlaylistTabsColumns = ({
  onDeleteClick,
}: UsePlaylistTabsColumnsProps) => {
  const skippedTrackColumns: Column[] = [
    {
      label: "",
      accessor: "image",
      primary: true,
      render: (image: Image) => (
        <div className="centered">
          <img alt="" width={40} height={40} src={image.url} />
        </div>
      ),
    },
    {
      label: "Title",
      accessor: "name",
      sortable: true,
      primary: true,
    },
    {
      label: "Artist",
      accessor: "artists",
      sortable: true,
      render: (artists: SkippedTrackArtist[]) => (
        <>
          {artists.map((artist, index) => (
            <>
              {`${index === 0 ? "" : ",,,,,"}`}
              {artist.name}
            </>
          ))}
        </>
      ),
    },
    {
      label: "Album",
      accessor: "album",
      sortable: true,
      render: (album: SkippedTrackAlbum) => <>{album.name}</>,
    },
    {
      label: <PiClock />,
      accessor: "duration",
      sortable: true,
      render: (millis: number) => {
        var minutes = Math.floor(millis / 60000);
        var seconds = parseInt(((millis % 60000) / 1000).toFixed(0));
        return seconds === 60
          ? minutes + 1 + ":00"
          : minutes + ":" + (seconds < 10 ? "0" : "") + seconds;
      },
    },
    {
      label: "Total Skipped",
      accessor: "skippedTotal",
      sortable: true,
      sortbyOrder: "desc",
    },
    {
      label: "",
      accessor: "id",
      render: (id: string) => (
        <div className="centered">
          <VscTrash
            style={{ cursor: "pointer" }}
            onClick={() => onDeleteClick(id)}
          />
        </div>
      ),
    },
  ];

  const skippedTrackHistoryColumns: Column[] = [
    {
      label: "",
      accessor: "image",
      primary: true,
      render: (image: Image) => (
        <div className="centered">
          <img alt="" width={40} height={40} src={image.url} />
        </div>
      ),
    },
    {
      label: "Title",
      accessor: "name",
      sortable: true,
      primary: true,
    },
    {
      label: "Artists",
      accessor: "artists",
      sortable: true,
      render: (artists: SkippedTrackArtist[]) => (
        <>
          {artists.map((artist) => (
            <>{artist.name}</>
          ))}
        </>
      ),
    },
    {
      label: "Album",
      accessor: "album",
      sortable: true,
      render: (album: SkippedTrackAlbum) => <>{album.name}</>,
    },
    {
      label: <PiClock />,
      accessor: "duration",
      sortable: true,
      render: (millis: number) => {
        var minutes = Math.floor(millis / 60000);
        var seconds = parseInt(((millis % 60000) / 1000).toFixed(0));
        return seconds === 60
          ? minutes + 1 + ":00"
          : minutes + ":" + (seconds < 10 ? "0" : "") + seconds;
      },
    },
    {
      label: "Date Skipped",
      accessor: "skippedDate",
      sortable: true,
      sortbyOrder: "desc",
      render: (date: Date) =>
        format(date, "MMM dd yyyy") + " at " + format(date, "HH:mm:ss"),
    },
    {
      label: "",
      accessor: "id",
      render: (id: string) => (
        <div className="centered">
          <VscTrash
            style={{ cursor: "pointer" }}
            onClick={() => onDeleteClick(id)}
          />
        </div>
      ),
    },
  ];

  const spotifyTrackColumns: Column[] = [
    {
      label: "",
      accessor: "image",
      primary: true,
      render: (image: Image) => (
        <div className="centered">
          <img alt="" width={40} height={40} src={image.url} />
        </div>
      ),
    },
    {
      label: "Title",
      accessor: "name",
      sortable: true,
      primary: true,
    },
    {
      label: "Artist",
      accessor: "artists",
      sortable: true,
      sortbyOrder: "asc",
      render: (artists: SkippedTrackArtist[]) => (
        <>
          {artists.map((artist) => (
            <>{artist.name}</>
          ))}
        </>
      ),
    },
    {
      label: "Album",
      accessor: "album",
      sortable: true,
      render: (album: SkippedTrackAlbum) => <>{album.name}</>,
    },
    {
      label: <PiClock />,
      accessor: "duration",
      sortable: true,
      render: (millis: number) => {
        var minutes = Math.floor(millis / 60000);
        var seconds = parseInt(((millis % 60000) / 1000).toFixed(0));
        return seconds === 60
          ? minutes + 1 + ":00"
          : minutes + ":" + (seconds < 10 ? "0" : "") + seconds;
      },
    },
    {
      label: "",
      accessor: "id",
      render: (id: string) => (
        <div className="centered">
          <VscTrash
            style={{ cursor: "pointer" }}
            onClick={() => onDeleteClick(id)}
          />
        </div>
      ),
    },
  ];

  return {
    skippedTrackColumns,
    skippedTrackHistoryColumns,
    spotifyTrackColumns,
  };
};
