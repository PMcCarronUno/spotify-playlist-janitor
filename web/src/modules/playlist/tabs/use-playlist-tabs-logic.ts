import { useCallback, useEffect, useState } from "react";
import { useDataApi } from "api/data-api";
import { useSpotifyApi } from "api/spotify-api";
import {
  DeleteTrack,
  SkippedTrackHistory,
  SkippedTrackTotal,
  Track,
} from "./playlist-tabs-types";
import { useModal } from "shared/components/modal";
import { usePlaylistTabsColumns } from "./use-playlist-tabs-columns";
import { Image } from "shared/types";
import { DEFAULT_ALBUM_IMAGE } from "shared/constants";

const loadingData: any[] = [{}, {}, {}, {}, {}, {}, {}, {}, {}];

type UsePlaylistTabsLogicProps = {
  id: string;
};

export const usePlaylistTabsLogic = ({ id }: UsePlaylistTabsLogicProps) => {
  const { getDatabasePlaylistSkippedTracks } = useDataApi();
  const { getSpotifyPlaylistTracks, deleteSpotifyPlaylistTracks } =
    useSpotifyApi();

  const [loadingSkippedTracks, setLoadingSkippedTracks] = useState(false);
  const [skippedTracks, setSkippedTracks] = useState<SkippedTrackTotal[]>([]);

  const [loadingSkippedTrackHistory, setLoadingSkippedTrackHistory] =
    useState(false);
  const [skippedTrackHistory, setSkippedTrackHistory] = useState<
    SkippedTrackHistory[]
  >([]);

  const [loadingSpotifyTracks, setLoadingSpotifyTracks] = useState(false);
  const [spotifyTracks, setSpotifyTracks] = useState<Track[]>([]);

  const { isOpen: deleteOpen, onOpen, onClose } = useModal();

  const [deleting, setDeleting] = useState(false);
  const [deleteError, setDeleteError] = useState<string | undefined>(undefined);
  const [deleteTracks, setDeleteTracks] = useState<DeleteTrack[]>([]);

  const onDeleteOpen = useCallback(
    (trackId: string) => {
      const skippedTrack = skippedTracks.find((track) => track.id === trackId);
      if (skippedTrack) {
        setDeleteTracks([{ id: skippedTrack?.id, name: skippedTrack?.name }]);
        onOpen();
      }
      const spotifyTrack = spotifyTracks.find((track) => track.id === trackId);
      if (spotifyTrack) {
        setDeleteTracks([{ id: spotifyTrack?.id, name: spotifyTrack?.name }]);
        onOpen();
      }
    },
    [skippedTracks, spotifyTracks, setDeleteTracks, onOpen]
  );

  const onDeleteClose = useCallback(() => {
    setDeleteTracks([]);
    setDeleteError(undefined);
    onClose();
  }, [setDeleteTracks, onClose]);

  const {
    skippedTrackColumns,
    skippedTrackHistoryColumns,
    spotifyTrackColumns,
  } = usePlaylistTabsColumns({ onDeleteClick: onDeleteOpen });

  const getSkippedTracks = useCallback(
    async (id: string) => {
      try {
        setLoadingSkippedTracks(true);
        setLoadingSkippedTrackHistory(true);
        setSkippedTracks([]);
        setSkippedTrackHistory([]);

        const skippedTrackResponse = await getDatabasePlaylistSkippedTracks(id);

        const skippedTracks: SkippedTrackTotal[] = [];
        var skippedTracksTotals: { [id: string]: SkippedTrackTotal } = {};

        skippedTrackResponse.data.forEach((skippedTrack) => {
          if (skippedTracksTotals[skippedTrack.trackId]) {
            skippedTracksTotals[skippedTrack.trackId].skippedTotal++;
          } else {
            skippedTracksTotals[skippedTrack.trackId] = {
              id: skippedTrack.trackId,
              name: skippedTrack.name,
              duration: skippedTrack.duration,
              skippedTotal: 1,
              album: skippedTrack.album,
              artists: skippedTrack.artists,
              image: skippedTrack.album.images.reduce((prev, curr) => {
                return prev.height < curr.height ? prev : curr;
              }),
            };
          }
        });

        for (let key in skippedTracksTotals) {
          skippedTracks.push(skippedTracksTotals[key]);
        }

        const skippedTrackHistory: SkippedTrackHistory[] =
          skippedTrackResponse.data.map((skippedTrack) => ({
            id: skippedTrack.trackId,
            name: skippedTrack.name,
            duration: skippedTrack.duration,
            skippedDate: new Date(skippedTrack.skippedDate),
            album: skippedTrack.album,
            artists: skippedTrack.artists,
            image: skippedTrack.album.images.reduce((prev, curr) => {
              return prev.height < curr.height ? prev : curr;
            }),
          }));

        setSkippedTracks(skippedTracks);
        setSkippedTrackHistory(skippedTrackHistory);

        setLoadingSkippedTracks(false);
        setLoadingSkippedTrackHistory(false);
      } catch (e: any) {
        // if (e?.response?.status === 404) {
        //   setNotFound(true);
        // }

        setSkippedTracks([]);
        setSkippedTrackHistory([]);

        setLoadingSkippedTracks(false);
        setLoadingSkippedTrackHistory(false);
      }
    },
    [
      setLoadingSkippedTracks,
      setLoadingSkippedTrackHistory,
      getDatabasePlaylistSkippedTracks,
    ]
  );

  const getSpotifyTracks = useCallback(
    async (id: string) => {
      try {
        setLoadingSpotifyTracks(true);
        setSpotifyTracks([]);

        const spotifyTrackResponse = await getSpotifyPlaylistTracks(id);

        const spotifyTracks: Track[] = spotifyTrackResponse.data.map(
          (spotifytrack) => {
            let image: Image = {
              height: 40,
              width: 40,
              url: DEFAULT_ALBUM_IMAGE,
            };

            if (spotifytrack.album.images.length > 0) {
              image = spotifytrack.album.images.reduce((prev, curr) => {
                return prev.height < curr.height ? prev : curr;
              });
            }

            return {
              id: spotifytrack.id,
              name: spotifytrack.name,
              duration: spotifytrack.duration,
              album: spotifytrack.album,
              artists: spotifytrack.artists,
              image,
            };
          }
        );

        setSpotifyTracks(spotifyTracks);

        setLoadingSpotifyTracks(false);
      } catch (e: any) {
        // if (e?.response?.status === 404) {
        //   setNotFound(true);
        // }
        setSpotifyTracks([]);
        setLoadingSpotifyTracks(false);
      }
    },
    [setLoadingSpotifyTracks, getSpotifyPlaylistTracks]
  );

  useEffect(() => {
    if (!!id) {
      getSkippedTracks(id);
      getSpotifyTracks(id);
    }
  }, [id, getSkippedTracks, getSpotifyTracks]);

  const onDeleteSubmit = useCallback(
    async (event: React.FormEvent<HTMLFormElement>) => {
      try {
        event.preventDefault();
        setDeleting(true);
        setDeleteError(undefined);
        const trackIds = deleteTracks.map((deleteTrack) => deleteTrack.id);
        await deleteSpotifyPlaylistTracks(id, trackIds);

        setDeleting(false);
        onDeleteClose();

        await getSkippedTracks(id);
        await getSpotifyTracks(id);
      } catch (e: any) {
        setDeleteError(e.response?.data?.message || "Unknown error");
        setDeleting(false);
      }
    },
    [
      id,
      setDeleting,
      setDeleteError,
      deleteTracks,
      deleteSpotifyPlaylistTracks,
      onDeleteClose,
      getSkippedTracks,
      getSpotifyTracks,
    ]
  );

  return {
    loadingData,
    loadingSkippedTracks,
    skippedTrackColumns,
    skippedTracks,
    loadingSkippedTrackHistory,
    skippedTrackHistoryColumns,
    skippedTrackHistory,
    loadingSpotifyTracks,
    spotifyTrackColumns,
    spotifyTracks,
    deleteTracks,
    deleteOpen,
    onDeleteClose,
    deleting,
    onDeleteSubmit,
    deleteError,
  };
};
