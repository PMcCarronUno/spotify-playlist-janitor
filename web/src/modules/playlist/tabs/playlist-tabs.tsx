import { Tab, TabList, TabPanel, Tabs } from "shared/components/tabs";
import { usePlaylistTabsLogic } from "./use-playlist-tabs-logic";
import { Table } from "shared/components/table";
import { skippedTrackColumns, skippedTrackHistoryColumns } from "./playlist-tabs-columns";
import { useMemo } from "react";

type PlaylistTabsProps = {
    id: string;
};

export const PlaylistTabs = ({ id }: PlaylistTabsProps) => {

    const { skippedTracks, skippedTrackHistory, loadingSkippedTracks, loadingSkippedTrackHistory } = usePlaylistTabsLogic({ id });

    const tabs = useMemo(() => <Tabs>
        <TabList>
            <Tab data-testid="skipped-tracks-tab">Skipped Tracks</Tab>
            <Tab data-testid="skipped-track-history-tab">Skipped Track History</Tab>
            <Tab data-testid="tracks-tab">Tracks</Tab>
        </TabList>

        <TabPanel>
            {!loadingSkippedTracks &&
                <Table
                    data={skippedTracks}
                    columns={skippedTrackColumns}
                />
            }
        </TabPanel>
        <TabPanel>
            {!loadingSkippedTrackHistory &&
                <Table
                    data={skippedTrackHistory}
                    columns={skippedTrackHistoryColumns}
                />
            }
        </TabPanel>
        <TabPanel>
            <>Tracks</>
        </TabPanel>
    </Tabs>, [loadingSkippedTracks, skippedTracks, loadingSkippedTrackHistory, skippedTrackHistory])

    return tabs;
}