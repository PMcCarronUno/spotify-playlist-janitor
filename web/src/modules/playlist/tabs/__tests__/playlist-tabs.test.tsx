import {
  fireEvent,
  getByTestId,
  queryByTestId,
  render,
  waitFor,
} from "@testing-library/react";

import { PlaylistTabs } from "../playlist-tabs";
import { usePlaylistTabsLogic } from "../use-playlist-tabs-logic";
import {
  playlist,
  skippedTrackTotals,
  skippedTrackHistory,
  spotifyTracks,
  imageColumn,
  albumColumn,
  artistColumn,
  dateSkippedColumn,
  durationColumn,
  getDeleteColumn,
  titleColumn,
  totalSkippedColumn,
} from "shared/mock-data/playlist";
import { usePlaylistTabsColumns } from "../use-playlist-tabs-columns";

jest.mock("../use-playlist-tabs-logic");


describe("<PlaylistTabs />", () => {
  let container: HTMLElement;
  let mockUsePlaylistTabsLogic: any;

  const {
    skippedTrackColumns,
    skippedTrackHistoryColumns,
    spotifyTrackColumns,
  } = usePlaylistTabsColumns({ onDeleteClick: (id: string): void => {} });

  beforeEach(async () => {
    mockUsePlaylistTabsLogic = {
      loadingData: [],
      loadingSkippedTracks: false,
      skippedTrackColumns,
      skippedTracks: skippedTrackTotals,
      loadingSkippedTrackHistory: false,
      skippedTrackHistoryColumns,
      skippedTrackHistory: skippedTrackHistory,
      loadingSpotifyTracks: false,
      spotifyTrackColumns,
      spotifyTracks: spotifyTracks,
      deleteTracks: [],
      deleteOpen: false,
      onDeleteClose: jest.fn(),
      deleting: false,
      onDeleteSubmit: jest.fn(),
      deleteError: undefined,
    };

    jest
      .mocked(usePlaylistTabsLogic)
      .mockImplementation(() => mockUsePlaylistTabsLogic);

    await waitFor(() => {
      ({ container } = render(
        <PlaylistTabs {...{ id: playlist.id, loading: false }} />
      ));
    });
  });

  it("should render PlaylistTabs component", () => {
    expect(container.firstChild).toMatchSnapshot();

    expect(
      container.getElementsByClassName("react-loading-skeleton").length
    ).toBe(0);
  });

  // it("should render Playlist component in loading state", async () => {
  //   mockUsePlaylistLogic = {
  //     ...mockUsePlaylistLogic,
  //     playlist: undefined,
  //     loading: true,
  //   };
  //   jest.mocked(usePlaylistLogic).mockImplementation(() => mockUsePlaylistLogic);

  //   await waitFor(() => {
  //     ({ container } = render(<Playlist {...{id: playlist.id}}/>));
  //   });

  //   expect(container.firstChild).toMatchSnapshot();

  //   expect(container.getElementsByClassName("react-loading-skeleton").length).toBe(5)
  // });

  // it("should render Home component with Not Found", async () => {
  //   mockUsePlaylistLogic = {
  //     ...mockUsePlaylistLogic,
  //     notFound: true,
  //   };
  //   jest.mocked(usePlaylistLogic).mockImplementation(() => mockUsePlaylistLogic);

  //   await waitFor(() => {
  //     ({ container } = render(<Playlist {...{id: playlist.id}}/>));
  //   });

  //   expect(container.firstChild).toMatchSnapshot();

  //   expect(getByTestId(container, "not-found")).toBeTruthy();
  // });
});
