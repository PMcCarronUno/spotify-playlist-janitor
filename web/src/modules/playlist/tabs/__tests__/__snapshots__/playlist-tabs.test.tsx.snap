// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlaylistTabs /> should render PlaylistTabs component 1`] = `
<div
  class="sc-beyTiQ kSNZbW"
  data-rttabs="true"
>
  <ul
    class="react-tabs__tab-list"
    role="tablist"
  >
    <li
      aria-controls="panel:r0:0"
      aria-disabled="false"
      aria-selected="true"
      class="react-tabs__tab react-tabs__tab--selected"
      data-rttab="true"
      data-testid="skipped-tracks-tab"
      id="tab:r0:0"
      role="tab"
      tabindex="0"
    >
      Skipped Tracks
    </li>
    <li
      aria-controls="panel:r0:1"
      aria-disabled="false"
      aria-selected="false"
      class="react-tabs__tab"
      data-rttab="true"
      data-testid="skipped-track-history-tab"
      id="tab:r0:1"
      role="tab"
    >
      Skipped Track History
    </li>
    <li
      aria-controls="panel:r0:2"
      aria-disabled="false"
      aria-selected="false"
      class="react-tabs__tab"
      data-rttab="true"
      data-testid="tracks-tab"
      id="tab:r0:2"
      role="tab"
    >
      Tracks
    </li>
  </ul>
  <div
    aria-labelledby="tab:r0:0"
    class="react-tabs__tab-panel react-tabs__tab-panel--selected"
    id="panel:r0:0"
    role="tabpanel"
  >
    No skipped tracks found.
  </div>
  <div
    aria-labelledby="tab:r0:1"
    class="react-tabs__tab-panel"
    id="panel:r0:1"
    role="tabpanel"
  />
  <div
    aria-labelledby="tab:r0:2"
    class="react-tabs__tab-panel"
    id="panel:r0:2"
    role="tabpanel"
  />
</div>
`;
